### PR TITLE
[dynamic-import-chunkname] Disallow chunk name when eager mode is enabled

### PR DESCRIPTION
**Pull request by amsardesai**
The dynamic-import-chunkname rule is useful for catching missing chunk names for imports, however much of our imports use `webpackMode: "eager"` to disable creating a separate chunk (i.e for lazy evaluation). Adding `webpackChunkName` anyway can force webpack to generate a chunk, defeating the purpose of the eager mode. This PR changes the rule to disallow chunk names when `webpackMode: "eager"` is set.

Related issue: https://github.com/import-js/eslint-plugin-import/issues/1904

### CODE REVIEW

**Comment by ljharb** [comment]
When would you not want this option enabled?

**Comment by amsardesai** [comment]
Never, really. I'll just make it the default.

**Comment by ljharb** [comment]
Is there anyone who this could break? I assume not, because the only complaint would be that it _fails_ to warn on something?

**Comment by amsardesai** [comment]
Exactly, this only suppresses existing warnings in a narrow case, and doesn't introduce new warnings.

**Review by ljharb — APPROVED** [review]
Thanks, this is great!

One more enhancement we may want to consider before landing this: when webpackMode is eager, and a useless name is present, could we give a suggestion to remove the name?

**Comment by amsardesai** [comment]
@ljharb Yeah that makes sense to me, just updated the PR to warn when chunk name and eager mode are both set.

File: src/rules/dynamic-import-chunkname.js
Comment (ljharb) [inline src/rules/dynamic-import-chunkname.js:112]:
```suggestion
      if (isChunknamePresent && isEagerModePresent) {
        context.report({
          node,
          message: 'dynamic imports should not have both webpackChunkName and webpackMode: "eager"`\',
        });
      }
```

also is there any reason why someone might still want a chunk name?

Comment (amsardesai) [reply src/rules/dynamic-import-chunkname.js:112]: I don't think having a chunk name with eager mode makes sense, since eager mode explicitly disables creating a new chunk, so the module is just lazily evaluated instead of lazily loaded+evaluated. Adding a chunk name anyways basically reverts back to being `webpackMode: 'lazy'` and generates a new chunk. A warning would help users catch unintentional additional chunks when using `webpackMode: 'eager'`

Comment (ljharb) [reply src/rules/dynamic-import-chunkname.js]: ok, so having both is always a mistake. thanks!

Comment (ljharb) [reply src/rules/dynamic-import-chunkname.js]: altho actually - that means that removing the name would quietly change the mode from lazy to eager - i think this should be a suggestion to *either* remove the name, *or* remove/change the webpackMode, to be safe.

**Comment by ljharb** [comment]
@amsardesai sorry if i wasn't clear; a "suggestion" isn't prose giving the user a hint, it's an actual eslint feature where the editor will prompt them with multiple autofixes they have to explicitly choose. For an example, look for a rule with `hasSuggestions: true` in its schema.

**Comment by amsardesai** [comment]
Ah, I didn't realize you meant fixers too. Let me do that.

File: src/rules/dynamic-import-chunkname.js
Comment (ljharb) [inline src/rules/dynamic-import-chunkname.js:40]: how come this one no longer has a leading space, but still has a trailing space?

Comment (amsardesai) [reply src/rules/dynamic-import-chunkname.js:40]: I'm using this regex in a `replace()` operation to remove it, and I didn't want the preceding space to be removed in a comma-separated list of webpack options, it could lead to the format being incorrect after applying the suggestion (i.e there could be no space after comma). The existence of the space is already checked by the lint rule using the regex right above this.

**Comment by amsardesai** [comment]
Hey @ljharb, would love to know if this can be merged now, or if any other changes need to be made!

**Comment by ljharb** [comment]
@amsardesai ok, i've rebased this and it's probably ready to go :-) just wondering about some of the remaining test _changes_ (not the additions, those make sense) - some are lazy → eager, and some are eager → lazy. Can you go through and comment in the PR on those lines, and explain what motivated the change? i suspect you have a convincing explanation for them all but this way it'll be documented for me in the future :-)

File: tests/src/rules/dynamic-import-chunkname.js
Comment (amsardesai) [inline tests/src/rules/dynamic-import-chunkname.js:358]: Removed the chunk name since it's not allowed anymore when eager mode is set

File: tests/src/rules/dynamic-import-chunkname.js
Comment (amsardesai) [inline tests/src/rules/dynamic-import-chunkname.js:415]: This test isn't checking for eager mode functionality, it's checking whether other modes are valid, so changing this to "lazy" so it doesn't interfere with the eager mode conditions.

File: tests/src/rules/dynamic-import-chunkname.js
Comment (amsardesai) [inline tests/src/rules/dynamic-import-chunkname.js:1272]: This test would fail if it was eager mode, changing it to lazy so it's more correct.

File: tests/src/rules/dynamic-import-chunkname.js
Comment (amsardesai) [inline tests/src/rules/dynamic-import-chunkname.js:1329]: Ditto here, this test is checking for webpack comment correctness, not eager mode, so changing this to "lazy"

**Review by amsardesai — COMMENTED** [review]
Also updated some tests to be a bit more thorough!

**Comment by amsardesai** [comment]
Hey @ljharb, would love to know when you will put out a new release - I'm hoping to use the changes in this PR in our repo.

**Comment by ljharb** [comment]
I was hoping to get flat config support landed before cutting a release; I'll plan to cut one within the next couple of weeks regardless.

**Comment by amsardesai** [comment]
Thanks for the update!

**Comment by amsardesai** [comment]
Hey @ljharb, just pinging again to check if you will release soon, since it's been 8 months since the last release.

**Comment by ljharb** [comment]
@amsardesai i'll probably have to cut a release without flat config support, sadly, so i'd expect one in the next few weeks.

**Comment by amsardesai** [comment]
thank you!

**Pull request merged**
